tools/dkcm/mainservice/mainservice.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mainservice
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	"knative.dev/test-infra/tools/dkcm/clerk"
+
+	"knative.dev/test-infra/pkg/mysql"
+)
+
+var (
+	dbConfig             *mysql.DBConfig
+	db                   *clerk.DB
+	DefaultClusterParams = clerk.ClusterParams{Zone: DefaultZone, Nodes: 4, NodeType: DefaultNodeType}
+)
+
+// handle cleaning cluster request after usage
+func handleCleanCluster(w http.ResponseWriter, r *http.Request) {
+
+}
+
+// check the pool capacity
+func checkPoolCap(cp *clerk.ClusterParams) {
+	numAvail := db.CheckNumAvail(cp)
+	if numAvail < DefaultOverProvision {
+		// create cluster if not meeting overprovisioning criteria
+		CreateCluster()
+	}
+}
+
+// assign clusters if available upon request
+func CreateCluster() {
+
+}
+
+// assign clusters if available upon request
+func AssignCluster(token string, w http.ResponseWriter) {
+	r, err := db.GetRequest(token)
+	if err != nil {
+		http.Error(w, "there is an error getting the request with the token. Please try again.", http.StatusNotAcceptable)
+		return
+	}
+	available, clusterID := db.CheckAvail(r.ClusterParams)
+	if available == true {
+		response, err := db.GetCluster(clusterID)
+		if err != nil {
+			http.Error(w, "there is an error getting available clusters. Please try again.", http.StatusInternalServerError)
+			return
+		}
+		db.UpdateRequest(r.RequestID, clerk.UpdateNumField("ClusterID", clusterID))
+		checkPoolCap(&DefaultClusterParams)
+		w.Write([]byte("Your cluster is ready! Here is the info:\n"))
```

Comment:
The response should be in JSON format rather than a raw string, something like
```
{
   isReady: false,
   message: "xxxxx"
   clusterInfo: {
      name: "xxx",
      project: "yyy",
      zone: "zzz"
   }
}
```

You can use the `encoding/json` library to do this.

Reply:
ah Chao told me it would suffice to let it be comma separated. Which way should I follow?